okpt/io/config/validator.py

```diff
@@ -0,0 +1,110 @@
+
+import cerberus
+
+from okpt.io import reader
+from okpt.io.config import parser
+
+_TOOL, _OPENSEARCH = ['tool', 'opensearch']
+
+
+class ConfigurationError(Exception):
+    """Exception raised for errors in the tool configuration.
+
+    Attributes:
+        message -- explanation of the error
+    """
+    def __init__(self, message: str):
+        self.message = message
+
+
+class _Validator():
+    """Validator for various configuration schemas.
+
+    Attributes:
+        validator: Cerberus validator for a particular schema
+        errors: Cerberus validation errors (if any are found during validation)
+    """
+    def __init__(self, schema_name: str):
+        self.validator = self._get_validator_from_schema_name(schema_name)
+        self.errors = ''
+
+    def _get_validator_from_schema_name(self, schema_name: str):
+        """Get the corresponding Cerberus validator from a schema name."""
+        curr_file_dir = os.path.dirname(os.path.abspath(__file__))
+        schemas_dir = os.path.join(curr_file_dir, 'schemas')
+        schema_file_path = os.path.join(schemas_dir, f'{schema_name}.yml')
+        schema_file_obj = reader.get_read_file_obj(schema_file_path)
+        schema_obj = parser.parse_yaml(schema_file_obj)
+        return cerberus.Validator(schema_obj)
+
+    def validate(self, config_obj: Dict[str, Any]):
+        """Validate the configuration obj against the class schema."""
+        if not self.validator.validate(config_obj):
+            self.errors = self.validator.errors
+            return False
+        return True
+
+
+def validate(tool_config_file_obj: TextIOWrapper) -> List[Any]:
+    """Validate the configurations of the tool.
+
+    Args:
+        tool_config_file_obj: File object of the tool configuration file.
+
+    Returns:
+        List with the tool configuration, service configuration, and index settings (if possible).
+
+    Raises:
+        ConfigurationError: If there is an error in the user configuration.
+    """
+
+    # validate tool config
+    tool_config_validator = _Validator(_TOOL)
+    tool_config_obj = parser.parse_yaml(tool_config_file_obj)
+    is_tool_config_valid = tool_config_validator.validate(tool_config_obj)
+
+    if not is_tool_config_valid:
+        raise ConfigurationError(
+            f'Tool Config Error: {tool_config_validator.errors}')
+
+    # validate service config
+    knn_service = tool_config_obj['knn_service']
+    service_config_validator = _Validator(knn_service)
+    service_config_file_path = tool_config_obj['service_config']
+    service_config_obj = parser.parse_yaml_from_path(
+        service_config_file_path)
+    is_service_config_valid = service_config_validator.validate(
+        service_config_obj)
+
+    if not is_service_config_valid:
+        raise ConfigurationError(
+            f'Service Config Error: {service_config_validator.errors}')
+
+    # check if index spec needs to be parsed (for opensearch only)
+    if knn_service == _OPENSEARCH:
```

Comment:
So my thinking here is that there should be a separate validator for each service. This way, we can just rely on line 97 to validate the services config. You could create some kind of factory method that, given the service name, return the correct validator.

Reply:
I've refactored the parsing/validation with the design that we discussed, using a `BaseParser` interface and creating config specific subclasses from it.